java/arcs/core/host/AbstractArcHost.kt

```diff
@@ -10,43 +10,239 @@
+                particleContext.run {
+                    particle.onHandleSync(handle, false)
+                }
+            }
+            particleContext.particleState = ParticleState.Started
+        }
+    }
+
+    suspend fun maybeRequestResurrection(context: ArcHostContext) {
+        if (context.arcState == ArcState.NeverStarted) {
+            // TODO: First time, request resurrection for context.
+        }
+    }
+
+    suspend fun maybeCancelResurrection(context: ArcHostContext) {
+        // TODO: wire up resurrection cancellation
+    }
+
+    /**
+     * Given a handle name, a [HandleConnection], and a [HandleHolder] construct an Entity
+     * [Handle] of the right type.
+     */
+    suspend fun createHandle(
+        handleName: String,
+        handleSpec: Plan.HandleConnection,
+        holder: HandleHolder
+    ) = when (handleSpec.type) {
+        is SingletonType<*> ->
+            entityHandleManager().createSingletonHandle(
+                holder,
+                handleName,
+                handleSpec.storageKey,
+                handleSpec.type.toSchema()
+            )
+        is CollectionType<*> ->
+            entityHandleManager().createSetHandle(
+                holder,
+                handleName,
+                handleSpec.storageKey,
+                handleSpec.type.toSchema()
+            )
+        else -> throw Exception("Unknown type ${handleSpec.type}")
     }
 
     override suspend fun stopArc(partition: Plan.Partition) {
-        // TODO: not implemented yet
+        val arcId = partition.arcId
+        val context = lookupOrCreateArcHostContext(partition)
+        runningArcs[partition.arcId]?.let { context ->
+            when (context.arcState) {
+                ArcState.Running -> stopArcInternal(arcId, context)
+                ArcState.NeverStarted -> stopArcError(context, "Arc $arcId was never started")
+                ArcState.Stopped -> stopArcError(context, "Arc $arcId already stopped")
+                ArcState.Deleted -> stopArcError(context, "Arc $arcId is deleted.")
+            }
+        }
+    }
+
+    suspend fun stopArcError(context: ArcHostContext, message: String) {
+        // TODO: decide how to propagate this
+    }
+
+    suspend fun stopArcInternal(arcId: String, context: ArcHostContext) {
+        for (particleContext in context.particles.values) {
+            stopParticle(particleContext)
+        }
+        maybeCancelResurrection(context)
+        context.arcState = ArcState.Stopped
+        updateArcHostContext(arcId, context)
+    }
+
+    suspend fun stopParticle(context: ParticleContext) {
+        context.particle.onShutdown()
+        context.particleState = ParticleState.Stopped
+        cleanupHandles(context.particle.handles)
+    }
+
+    suspend fun cleanupHandles(handles: HandleHolder) {
+        for ((name, handle) in handles.handles) {
+            // TODO: disconnect/unregister handle
+        }
+        (handles.handles as MutableMap).clear()
```

Comment:
Is this forcing something exposed as a `Map` to be a `MutableMap`? Seems fishy/dangerous and like it's making the AbstractArcHost know too much about the internals of HandleHolder.

Reply:
HandleHolder is kind of the glue between Particle and ArcHost. The Particle sees typed fields, the ArcHost sees a map. The HandleHolder is statically code-genned at built-time.

I could make the HandleHolder field a MutableMap, but I think Cameron or Piotr once objected to this IIRC. We kind of wanted the Handle fields to be immutable from the view of the Particle. 

However, I'm okay with just making it mutable.

Reply:
Rather than making it mutable, could you add a `.clear()` or some appropriately-named method to the class that owns the map?

Reply:
Done